From 9a527d983b77862e69484265d27b1f618f1517af Mon Sep 17 00:00:00 2001
From: Christian Hewitt <christianshewitt@gmail.com>
Date: Tue, 31 Mar 2026 07:42:58 +0000
Subject: [PATCH 13/23] media: meson: vdec: esparser: move global state to
 per-core structure

The esparser uses a module-level static wait queue and search_done flag
shared across all potential instances of the driver. If multiple decoder
cores were ever instantiated, one core's ISR would wake the other
core's waiter, causing incorrect operation.

Move the wait queue and completion flag into struct amvdec_core so each
hardware instance has its own synchronization state. Initialize the
wait queue in esparser_init().

Signed-off-by: Christian Hewitt <christianshewitt@gmail.com>
---
 drivers/staging/media/meson/vdec/esparser.c | 14 +++++++-------
 drivers/staging/media/meson/vdec/vdec.h     |  4 ++++
 2 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/drivers/staging/media/meson/vdec/esparser.c b/drivers/staging/media/meson/vdec/esparser.c
index edfe319b8fa10..bf401aa12205a 100644
--- a/drivers/staging/media/meson/vdec/esparser.c
+++ b/drivers/staging/media/meson/vdec/esparser.c
@@ -53,9 +53,6 @@
 #define SEARCH_PATTERN_LEN	512
 #define VP9_HEADER_SIZE		16
 
-static DECLARE_WAIT_QUEUE_HEAD(wq);
-static int search_done;
-
 static irqreturn_t esparser_isr(int irq, void *dev)
 {
 	int int_status;
@@ -67,8 +64,8 @@ static irqreturn_t esparser_isr(int irq, void *dev)
 	if (int_status & PARSER_INTSTAT_SC_FOUND) {
 		amvdec_write_parser(core, PFIFO_RD_PTR, 0);
 		amvdec_write_parser(core, PFIFO_WR_PTR, 0);
-		search_done = 1;
-		wake_up_interruptible(&wq);
+		core->esparser_done = 1;
+		wake_up_interruptible(&core->esparser_wq);
 	}
 
 	return IRQ_HANDLED;
@@ -228,8 +225,9 @@ esparser_write_data(struct amvdec_core *core, dma_addr_t addr, u32 size)
 			    (7 << FETCH_ENDIAN_BIT) |
 			    (size + SEARCH_PATTERN_LEN));
 
-	search_done = 0;
-	return wait_event_interruptible_timeout(wq, search_done, (HZ / 5));
+	core->esparser_done = 0;
+	return wait_event_interruptible_timeout(core->esparser_wq,
+					       core->esparser_done, (HZ / 5));
 }
 
 static u32 esparser_vififo_get_free_space(struct amvdec_session *sess)
@@ -450,6 +448,8 @@ int esparser_init(struct platform_device *pdev, struct amvdec_core *core)
 	int ret;
 	int irq;
 
+	init_waitqueue_head(&core->esparser_wq);
+
 	irq = platform_get_irq_byname(pdev, "esparser");
 	if (irq < 0)
 		return irq;
diff --git a/drivers/staging/media/meson/vdec/vdec.h b/drivers/staging/media/meson/vdec/vdec.h
index f6ca1ba146f35..ec8b259fe22a6 100644
--- a/drivers/staging/media/meson/vdec/vdec.h
+++ b/drivers/staging/media/meson/vdec/vdec.h
@@ -62,6 +62,8 @@ struct amvdec_session;
  * @vdec_hevc_clk: VDEC_HEVC clock
  * @vdec_hevcf_clk: VDEC_HEVCF clock
  * @esparser_reset: RESET for the PARSER
+ * @esparser_wq: wait queue for ESPARSER completion
+ * @esparser_done: flag signaled by ESPARSER ISR on completion
  * @vdev_dec: video device for the decoder
  * @v4l2_dev: v4l2 device
  * @m2m_dev: v4l2 m2m device
@@ -86,6 +88,8 @@ struct amvdec_core {
 	struct clk *vdec_hevcf_clk;
 
 	struct reset_control *esparser_reset;
+	wait_queue_head_t esparser_wq;
+	int esparser_done;
 
 	struct video_device *vdev_dec;
 	struct v4l2_device v4l2_dev;
